### PR TITLE
Prepare all active version support

### PR DESCRIPTION
**Pull request by Ralgathor**
@mrbuds @hypernormalisation if you have time I would like a review from you on this changes to allow support for all active version of the game.

You can ignore the "[Split table data on different files by version](https://github.com/Ralgathor/LibClassicSwingTimerAPI/pull/29/commits/f232f34876fbecece41a2128a316f4c04225aa08)" commit as I have reverted this change to keep the lib to one single file.

### CODE REVIEW

File: SwingTimerAPI.lua [hunk 1]
```diff
@@ -9,152 +9,174 @@ local C_Timer, tonumber = C_Timer, tonumber
 local GetSpellInfo, GetTime, CombatLogGetCurrentEventInfo = GetSpellInfo, GetTime, CombatLogGetCurrentEventInfo
 local UnitAttackSpeed, UnitAura, UnitGUID, UnitRangedDamage = UnitAttackSpeed, UnitAura, UnitGUID, UnitRangedDamage
 
+local isRetails = WOW_PROJECT_ID == WOW_PROJECT_MAINLINE
```
Comment (mrbuds) [inline SwingTimerAPI.lua:12]: typo => isRetail

File: SwingTimerAPI.lua [hunk 2]
```diff
@@ -9,152 +9,174 @@ local C_Timer, tonumber = C_Timer, tonumber
...
+local isClassic = WOW_PROJECT_ID == WOW_PROJECT_CLASSIC
+local isBCC = WOW_PROJECT_ID == WOW_PROJECT_BURNING_CRUSADE_CLASSIC and LE_EXPANSION_LEVEL_CURRENT == LE_EXPANSION_BURNING_CRUSADE
+local isWarth = WOW_PROJECT_ID == WOW_PROJECT_BURNING_CRUSADE_CLASSIC and LE_EXPANSION_LEVEL_CURRENT == LE_EXPANSION_WRATH_OF_THE_LICH_KING
```
Comment (mrbuds) [inline SwingTimerAPI.lua:15]: typo => isWrath

File: SwingTimerAPI.lua [hunk 3]
```diff
@@ -275,8 +298,13 @@ function lib:SwingEnd(hand)
 	self:Fire("SWING_TIMER_STOP", hand)
 	if (self.casting or self.channeling) and self.isAttacking and hand ~= "ranged" then
 		local now = GetTime()
-		self:SwingStart(hand, now, true)
-		self:Fire("SWING_TIMER_CLIPPED", hand)
+		if isRetails and hand == "mainhand" then		
+			self:SwingStart(hand, now, true)
+			self:Fire("SWING_TIMER_CLIPPED", hand)
+		elseif not isRetails then
```
Comment (mrbuds) [inline SwingTimerAPI.lua:304]: avoid negative tests for game version, better be explicit on which version code block should run

File: SwingTimerAPI.lua [hunk 4]
```diff
@@ -315,11 +343,15 @@ function lib:COMBAT_LOG_EVENT_UNFILTERED(_, ts, subEvent, _, sourceGUID, _, _, _
 		if isOffHand then
 			self.firstOffSwing = true
 			self:SwingStart("offhand", now, false)
-			self:SwingStart("ranged", now, true)
+			if not isClassicOrBCC and not isRetails then
```
Comment (mrbuds) [inline SwingTimerAPI.lua:346]: here too

File: SwingTimerAPI.lua [hunk 5]
```diff
@@ -315,11 +343,15 @@ function lib:COMBAT_LOG_EVENT_UNFILTERED(_, ts, subEvent, _, sourceGUID, _, _, _
...
+				self:SwingStart("ranged", now, true)
+			end
 		else
 			self.firstMainSwing = true
 			self:SwingStart("mainhand", now, false)
-			self:SwingStart("ranged", now, true)
+			if not isClassicOrBCC and not isRetails then
```
Comment (mrbuds) [inline SwingTimerAPI.lua:352]: and here

File: SwingTimerAPI.lua [hunk 6]
```diff
@@ -436,16 +472,22 @@ end
 
 function lib:UNIT_SPELLCAST_SUCCEEDED(_, _, _, spell)
 	local now = GetTime()
-	if spell ~= nil and next_melee_spells[spell] then
+	if spell ~= nil and next_melee_spells[spell] and next_melee_spells[spell][gameVersion] then
 		self:SwingStart("mainhand", now, false)
-		self:SwingStart("ranged", now, true)
+		if not self.isClassicOrBCC and not isRetails then
```
Comment (mrbuds) [inline SwingTimerAPI.lua:477]: here

File: SwingTimerAPI.lua [hunk 7]
```diff
@@ -472,12 +515,14 @@ function lib:UNIT_SPELLCAST_SUCCEEDED(_, _, _, spell)
 			if enabled == 1 then -- Reset ranged swing when FD CD start
 				self:SwingStart("mainhand", start, true)
 				self:SwingStart("offhand", start, true)
-				self:SwingStart("ranged", start, true)
+				if not isRetails then
```
Comment (mrbuds) [inline SwingTimerAPI.lua:518]: here

File: SwingTimerAPI.lua [hunk 8]
```diff
@@ -509,19 +554,32 @@ end
...
+	end
 end
 
 function lib:PLAYER_EQUIPMENT_CHANGED(_, equipmentSlot)
 	if equipmentSlot == 16 or equipmentSlot == 17 or equipmentSlot == 18 then
 		local now = GetTime()
 		self:SwingStart("mainhand", now, true)
 		self:SwingStart("offhand", now, true)
-		self:SwingStart("ranged", now, true)
+		if not isRetails then
```
Comment (mrbuds) [inline SwingTimerAPI.lua:580]: and here

**Review by mrbuds — COMMENTED** [review]
> @mrbuds @hypernormalisation if you have time I would like a review from you on this changes to allow support for all active version of the game.
> 
> You can ignore the "[Split table data on different files by version](https://github.com/Ralgathor/LibClassicSwingTimerAPI/pull/29/commits/f232f34876fbecece41a2128a316f4c04225aa08)" commit as I have reverted this change to keep the lib to one single file.

Having spell tables per game extension is healthy, easier to maintain, use less memory and reduce amount of table lookup to access values
I'd consider revert the revert

File: SwingTimerAPI.lua [hunk 9]
```diff
@@ -198,11 +231,66 @@ function lib:ADDON_LOADED(_, addOnName)
...
 	self.isAttacking = false
 	self.preventSwingReset = false
+	self.auraPreventSwingReset = false
 	self.skipNextAttack = nil
 	self.skipNextAttackCount = 0
 
 	self.skipNextAttackSpeedUpdate = nil
 	self.skipNextAttackSpeedUpdateCount = 0
+
+	for spell in pairs(list_reset_swing_spells) do
```
Comment (Ralgathor) [inline SwingTimerAPI.lua:241]: Dynamiclly construct the table on ADDON_LOADED

File: SwingTimerAPI.lua [hunk 10]
```diff
@@ -227,6 +237,60 @@ function lib:ADDON_LOADED(_, addOnName)
...
+		if list_reset_swing_spells[spell][gameVersion] then
+			reset_swing_spells[spell] = true
+		end
+	end
```
Comment (Ralgathor) [inline SwingTimerAPI.lua:245]: Dynamiclly construct the table on ADDON_LOADED

Comment (Ralgathor) [inline SwingTimerAPI.lua:245]: Dynamiclly construct the table on ADDON_LOADED

**Comment by mrbuds** [comment]
> My goal it's to maintain the lib in one unique file to be more easly embeded in another addons

Indeed i was baited by toc files and forget addons use xml .. but you could make a xml file per game version and it's up to addon to load correct xml file

Otherwise if you keep it in one file i think it's easier and more efficient to do this
```Lua
local list_reset_swing_spells
local reset_swing_channel_spells
local prevent_swing_speed_update
if isClassic then
  list_reset_swing_spells = {
    ...
  }
  reset_swing_channel_spells = {
    ...
  }
  prevent_swing_speed_update= {
    ...
  }
elseif IsBCC then
  list_reset_swing_spells = {
    ...
  }
  reset_swing_channel_spells = {
    ...
  }
  prevent_swing_speed_update= {
    ...
  }
elseif IsWrath then

elseif IsRetail then

end
```
Current way memory hold tables with data for all versions, use them only 1 time, then make more tables

**Comment by Ralgathor** [comment]
> > My goal it's to maintain the lib in one unique file to be more easly embeded in another addons
> 
> Indeed i was baited by toc files and forget addons use xml .. but you could make a xml file per game version and it's up to addon to load correct xml file
> 
> Otherwise if you keep it in one file i think it's easier and more efficient to do this
> 
> ```lua
> local list_reset_swing_spells
> local reset_swing_channel_spells
> local prevent_swing_speed_update
> if isClassic then
>   list_reset_swing_spells = {
>     ...
>   }
>   reset_swing_channel_spells = {
>     ...
>   }
>   prevent_swing_speed_update= {
>     ...
>   }
> elseif IsBCC then
>   list_reset_swing_spells = {
>     ...
>   }
>   reset_swing_channel_spells = {
>     ...
>   }
>   prevent_swing_speed_update= {
>     ...
>   }
> elseif IsWrath then
> 
> elseif IsRetail then
> 
> end
> ```
> 
> Current way memory hold tables with data for all versions, use them only 1 time, then make more tables Yes there will be duplicate lines but it's better have duplicates lines in code than unused data in memory

I see the issue to having table that are only used one time. Can we make this table garbage collected? I read some stuff one Lua memory usage and it seems if I set this tables local ref to nil the table should normally be garbage collected.

I'm really not fan of maintaining multiple tables by game version. It's seem very messy for me and I prefer the current format.

**Comment by Ralgathor** [comment]
> I see the issue to having table that are only used one time. Can we make this tables garbage collected? I read some stuff one Lua memory usage and it seems if I set this tables local ref to nil the table should normally be garbage collected.
> 
> I'm really not fan of maintaining multiple tables by game version. It's seem very messy for me and I prefer the current format.

Nevermind, I read more about Lua code perf and it's better to create table with already define size that doing this dynamic table construct. I think that I need to accept for performance to have a more messy code to maintain!

**Pull request merged**